### PR TITLE
Teams/add past teams

### PR DESCRIPTION
**Pull request by juuh42dias**
This PR is related to the issue  #502
- Include select dropdown to filter teams by year

## Select teams by year
screenshot from 2017-08-11 17-12-23

## Select teams by year and order by activities
screenshot from 2017-08-11 17-12-35

### CODE REVIEW

File: app/controllers/teams_controller.rb [hunk 1]
```diff
@@ -106,4 +101,23 @@ def set_display_roles
       @display_roles = ['student']
       @display_roles.map!(&:pluralize)
     end
+
+    def show_teams_by(params)
```
Comment (carpodaster) [inline app/controllers/teams_controller.rb]: `params` is available as a controller method, you don't need to pass it as an argument. With the argument gone, I would go for a more generic method name, something like `find_teams`

It is also a veeeery long method.

I'm wondering why we only include the activities (w/ sorting) when there is a `:sort` param. I would think this way of sorting / eager loading makes sense whether or not we sort stuff. Do you know, @klappradla or @alicetragedy ?

Provided we allow combining all these options, we can just build up the query instead of having a complex `case` or `if … else` statement:

```ruby
direction = params[:direction] == 'asc' ? 'ASC' : 'DESC'
# start with the "base scope"
@teams = Team.by_season_phase.includes(:activities).order("teams.kind, activities.created_at #{direction}").references(:activities)
# Re-assign @teams if conditions apply:
@teams = @teams.by_season if foo
@teams = @teams.by_something_else if bar
```

Does that make sense?

File: app/models/team.rb [hunk 2]
```diff
@@ -63,10 +63,21 @@ def in_current_season
       where(season: Season.current)
     end
 
+    def by_season_year(year)
+      where(season: Season.where(name: year))
+    end
+
+    def select_teams_by_season_year(year)
```
Comment (carpodaster) [inline app/models/team.rb]: This looks like a highly specialised method. Since the `by_season_year` returns an active relation, you could just add `.selected` after that. Let us go that route and remove this method.

Or if that doesn't work, try rephrasing the `selected` method as a `scope`.

File: app/models/team.rb [hunk 3]
```diff
@@ -63,10 +63,21 @@ def in_current_season
...
+      by_season_year(year).selected
+    end
+
     def selected
       where(kind: %w(sponsored voluntary))
     end
 
+    def activities_ordered_by(direction)
```
Comment (carpodaster) [inline app/models/team.rb]: This, too, seems like a highly specialised method and to me it doesn't feel like ordering the activities is part of the `Team` class' responsibility. I would move it back into controller.

File: app/models/team.rb [hunk 4]
```diff
@@ -63,10 +63,21 @@ def in_current_season
       where(season: Season.current)
     end
 
+    def by_season_year(year)
```
Comment (carpodaster) [inline app/models/team.rb]: Idea: making this `by_season(year_or_season)` that automatically detects whether the argument is a real `Season` object or a year (either as an integer or a string) and then constructs the `where` condition accordingly.

File: app/assets/javascripts/past_teams.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+$(function() {
+  var $selectedYear = $('select#past_teams');
+
+  $selectedYear.on('change', function(){
+    var yearParam = "?year="+this.value;
+    var newUrl;
+    var removed = window.location.search;
+
+    if(removed.length>0){
```
Comment (carpodaster) [inline app/assets/javascripts/past_teams.js]: I cannot wrap my head around how that would behave in the browser I'm afraid. Can someone else check that pls? 😁

**Comment by klappradla** [comment]
Hey @juuh42dias looking at your screenshot I just realized you're using a native dropdown element _(looks a bit bootstrappy on Linux on first sight)_

However, in Chrome on Mac, this would end up looking like this:
screen shot 2017-08-13 at 17 19 53


The Bootstrap - and probably preferred way - would be this:
screen shot 2017-08-13 at 17 25 51

The changes for this would just be:
```ruby
= select_tag 'past_teams', options_for_select(Season.all_years), prompt: 'Past Teams', class: 'form-control'
```

However, since you'r not adding a form on this page, I think the element you're looking for is actually [a button with a dropdown](https://getbootstrap.com/docs/4.0/components/dropdowns/).

File: app/assets/javascripts/past_teams.js (hunk 5 above)
Comment (klappradla) [inline app/assets/javascripts/past_teams.js]: What it does: takes the current URL, removes the search params from it, and then appends new ones, e.g. `?year="2016"` to it.

It then changes the current URL which makes the browser do another get request with the new params.

Afaik @juuh42dias, you could shorten this to just:
```javascript
$(function() {
  var $selectedYear = $('select#past_teams');

  $selectedYear.on('change', function(){
    var yearParam = "?year="+this.value;
    window.location.search = yearParam;
  });
});
```
It should to exactly the same.

File: app/assets/javascripts/past_teams.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+$(function() {
+  var $selectedYear = $('select#past_teams');
```
Comment (klappradla) [inline app/assets/javascripts/past_teams.js]: This adds a lot of non-default behavior to this select element. My personal guideline for this is usually: if I use Javascript to add behavior to an element, use data attributes, in this case something like`data-behavior="switch-teams-year"` and leave classes and ids fully for anything related to CSS and styling.

I even added some JS code to the Teams App once where I use it. Have a look at [city.js](https://github.com/rails-girls-summer-of-code/rgsoc-teams/blob/master/app/assets/javascripts/city_fields.js) and the view files associated with it for inspiration.

Aside from that: I think it's sematically speaking slightly wrong to use a `select` tag here, since we're not in a form... What do you think?

File: app/controllers/teams_controller.rb (hunk 1 above)
Comment (klappradla) [inline app/controllers/teams_controller.rb]: 👍 sure, the eager loading should be independent of whether there's a `:sort` param or not - the `last-activity` timestamp is still displayed. Don't see the reason why it should be different @carpodaster.

Why we need to eager load all activities to then in the end just pluck a single timestamp per team is a different thing though 😉

File: app/assets/javascripts/past_teams.js (hunk 6 above)
Comment (branquinhoaa) [inline app/assets/javascripts/past_teams.js]: Really nice using data attributes for it!! 
I looked in your city.js but I couldn't find any view associated with it. I searched in the files and I put a breakpoint in your function to see the element. Is it used to auto-complete cities' names in forms? 
Anyway, I will replace the id by the data attribute.

Comment (klappradla) [inline app/assets/javascripts/past_teams.js]: Yes, you're right, it's used by this type of input element: [here's the code for it](https://github.com/rails-girls-summer-of-code/rgsoc-teams/blob/master/app/inputs/city_input.rb)

It's not 100% the same, but the usage in views is similar in syntax.

**Comment by juuh42dias** [comment]
@klappradla  @carpodaster  that's ok now?

File: app/controllers/teams_controller.rb [hunk 7]
```diff
@@ -6,11 +6,19 @@ class TeamsController < ApplicationController
   load_and_authorize_resource except: [:index, :show]
 
   def index
```
Comment (klappradla) [inline app/controllers/teams_controller.rb]: This looks a bit confusing to me 🙈 
What's the point in 3 times overriding the `teams` instance variable`? This is going to end up in multiple - redundant after all - calls to the database, which is usually not what you want. Can you think of a way to structure this nicer and get rid of redundant DB calls?

File: app/models/team.rb [hunk 8]
```diff
@@ -43,6 +43,10 @@ class Team < ActiveRecord::Base
 
   scope :accepted, -> { where(kind: %w(sponsored voluntary)) }
 
+  scope :by_season_year, ->(year) { joins(:season).where('seasons.name': year) }
+
+  scope :with_last_activities_ordered_by, ->(direction) { includes(:activities).order("teams.kind, activities.created_at #{direction}").references(:activities) }
```
Comment (klappradla) [inline app/models/team.rb]: I don't see this scope being used anywhere _(nor is it tested)_...

File: app/models/team.rb [hunk 9]
```diff
@@ -70,6 +74,18 @@ def selected
       where(kind: %w(sponsored voluntary))
     end
 
+    def by_season(year_or_season)
```
Comment (klappradla) [inline app/models/team.rb]: For what reason was this introduced? I don't see it being used anywhere 🤔

File: spec/controllers/teams_controller_spec.rb [hunk 10]
```diff
@@ -41,6 +41,13 @@
       end
     end
 
+    context 'past teams' do
+      it 'passing year' do
```
Comment (klappradla) [inline spec/controllers/teams_controller_spec.rb]: Hm, I'm afraid this does not say anything... Aside from the description not really telling the reader what's supposed to happen, the test only tests if the `index` action renders the `index.html` template. One can pass arbitrary params to an `index` action. The test would be the same with:
```ruby
get :index, params: { foo: 'bar'}
```

Given that the `index` action is rather complex, this test should probably better target the actual behavior there.

File: app/assets/javascripts/conferences_options.js [hunk 11]
```diff
@@ -1,20 +0,0 @@
-$(function() {
```
Comment (juuh42dias) [inline app/assets/javascripts/conferences_options.js]: We don't use this `conferences_options.js` anymore.

File: app/controllers/teams_controller.rb (hunk 7 above)
Comment (juuh42dias) [inline app/controllers/teams_controller.rb]: I accept suggestions for refactoring this function.

My brain is blocked for now. :crying_cat_face:

File: app/assets/javascripts/conferences_options.js (hunk 11 above)
Comment (klappradla) [inline app/assets/javascripts/conferences_options.js]: Ah ok, so it was just forgotten to be removed in another PR?

File: app/models/team.rb (hunk 9 above)
Comment (juuh42dias) [inline app/models/team.rb]: It was a Carsten suggest.

File: app/assets/javascripts/conferences_options.js (hunk 11 above)
Comment (juuh42dias) [inline app/assets/javascripts/conferences_options.js]: Unfortunately :disappointed:

Comment (klappradla) [inline app/assets/javascripts/conferences_options.js]: Oh, alright, thanx for clearifying 🙏

File: app/models/team.rb (hunk 9 above)
Comment (klappradla) [inline app/models/team.rb]: Ok, I see. But what usecase did he intend it for?

Maybe I'm getting things wrong here, but I don't see the point in adding new _unused_ methods - that's like introducing legacy on purpose 😉

File: app/controllers/teams_controller.rb (hunk 7 above)
Comment (klappradla) [inline app/controllers/teams_controller.rb]: Oh, no worries 💚 

First of all I'd try to get rid of redundant queries. So a super simple thing here could for instance be to move the conditional:
```ruby
def index
  # stuff
  if year.present?
    @teams = query1
  else
    @teams = query2
  end
end
```

Then move the two queries from the second assignment into one:
```ruby
@teams = Team
  .by_season(year)
  .accepted
  .includes(:activities)
  .order("teams.kind, activities.created_at #{direction}")
  .references(:activities)
```

Can you use this as a starting point?

Comment (juuh42dias) [reply app/controllers/teams_controller.rb]: @klappradla done :)

File: spec/controllers/teams_controller_spec.rb [hunk 12]
```diff
@@ -41,6 +41,24 @@
...
+    context 'past teams' do
+      it 'return index by season year' do
+        get :index, params: { year: '2016'}
+        expect(response).to render_template 'index'
+      end
+
+      it 'return teams from current season' do
+        create_list :team, 2, season: Season.current
+        get :index, params: { year: Season.current.name }
+        expect(assigns(:teams).count).to eq 2
```
Comment (klappradla) [inline spec/controllers/teams_controller_spec.rb]: How about making this even more strict?
```ruby
teams = create_list :team, 2, season: Season.current
expect(assigns(:teams)).to match_array(teams)
```

File: spec/controllers/teams_controller_spec.rb [hunk 13]
```diff
@@ -41,6 +41,24 @@
...
+      end
+
+      it 'return empty list when the season is empty of teams' do
+        get :index, params: { year: '2000' }
+        expect(assigns(:teams)).to match_array []
```
Comment (klappradla) [inline spec/controllers/teams_controller_spec.rb]: Hm, not sure if this test is really necessary. Since there aren't any `Team` records in the database, this would be positive for basically any query on the `team` table. I would suggest to either completely skip it, or add some records to really prove this one case.

File: spec/models/team_spec.rb [hunk 14]
```diff
@@ -211,6 +211,21 @@
       end
     end
 
+    describe '.by_season' do
```
Comment (klappradla) [inline spec/models/team_spec.rb]: I would suggest to make this a bit more explicit:
```ruby
let(:season) { create :season, name: '2016' }
let(:teams2016) { create_list :team, 2, season: season }

before { # create a team in different season than 2016 }

it "does something" do
  expect(Team.by_season('2016')).to match_array(teams2016))
end
```

What do you think?

File: spec/controllers/teams_controller_spec.rb (hunk 13 above)
Comment (klappradla) [reply spec/controllers/teams_controller_spec.rb]: Any news on what I wrote ☝️ @juuh42dias ?

One option could also be: move setting up the teams into a `let` macro
```ruby
let!(:teams) { create_list :team, 2, season: Season.current }
```

Doing so, the test around year 2000 is a bit more meaningful - what do you think?

**Comment by klappradla** [comment]
🚢
